compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -33,6 +33,37 @@ impl<'tcx> MirPass<'tcx> for InstCombine {
             // Then carry out those optimizations.
             MutVisitor::visit_body(&mut InstCombineVisitor { optimizations, tcx }, body);
         }
+        remove_zst_assignments(tcx, body);
+    }
+}
+
+/// Assignments to ZST locals can be eliminated
+fn remove_zst_assignments<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    let def_id = body.source.def_id();
+    if tcx.is_static(def_id) {
+        return;
+    }
+    let param_env = tcx.param_env(def_id);
+
+    let local_decls = body.local_decls.clone();
+    for bbds in body.basic_blocks_mut() {
+        bbds.statements.retain(|stmt| {
+            match stmt.kind {
+                StatementKind::Assign(box (lhs, _)) => {
+                    let ty = local_decls[lhs.local].ty;
+                    let param_env_and = param_env.and(ty);
+                    if let Ok(layout) = tcx.layout_of(param_env_and) {
+                        if layout.is_zst() {
+                            debug!("Removing assign: {:?}", stmt);
+                            return false;
+                        }
+                    }
+                }
+                _ => {}
+            };
+
+            return true;
```

Comment:
```suggestion
            }

            true
```

one more nit, then r=me